### PR TITLE
Feature/auto instrumentation/cake php

### PR DESCRIPTION
**Pull request by AA2512**
In this I have added the instrumentation for Cake PHP, where I have hooked invokeAction function which is called whenever a request to any controller is made

### CODE REVIEW

File: src/Instrumentation/Slim/src/SlimInstrumentation.php [hunk 1]
```diff
@@ -68,8 +68,8 @@ public static function register(): void
                     $span = $builder->startSpan();
                 }
                 Context::storage()->attach($span->storeInContext($parent));
-
-                return [$request];
+                // return;
+                return $params;
```
Comment (ChrisLightfootWild) [inline src/Instrumentation/Slim/src/SlimInstrumentation.php:72]: Accidental commit?

File: src/Instrumentation/CakePHP/src/CakePHPInstrumentation.php [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                }
+                $scope->detach();
+                $span = Span::fromContext($scope->context());
+                $response = $app->getResponse();
+                if ($exception) {
+                    $span->recordException($exception, [TraceAttributes::EXCEPTION_ESCAPED => true]);
+                    $span->setStatus(StatusCode::STATUS_ERROR, $exception->getMessage());
+                }
+                if ($response) {
+                    error_log();
```
Comment (ChrisLightfootWild) [inline src/Instrumentation/CakePHP/src/CakePHPInstrumentation.php:75]: :scissors: 
```suggestion
```

File: src/Instrumentation/CakePHP/tests/Integration/CakeInstrumentationTest.php
Comment (ChrisLightfootWild) [inline src/Instrumentation/CakePHP/tests/Integration/CakeInstrumentationTest.php:1]: Is it worth adding some simple tests as a starting point? :thinking:

**Review by ChrisLightfootWild — CHANGES_REQUESTED** [review]

File: src/Instrumentation/Slim/src/SlimInstrumentation.php (hunk 1 above)
Comment (AA2512) [reply src/Instrumentation/Slim/src/SlimInstrumentation.php:72]: Yeah, will remove it

File: src/Instrumentation/CakePHP/src/CakePHPInstrumentation.php (hunk 2 above)
Comment (AA2512) [reply src/Instrumentation/CakePHP/src/CakePHPInstrumentation.php:75]: Sure

File: src/Instrumentation/CakePHP/tests/Integration/CakeInstrumentationTest.php
Comment (AA2512) [reply src/Instrumentation/CakePHP/tests/Integration/CakeInstrumentationTest.php:1]: Actually I thought that it was necessary to add tests. But I haven't done any testing before in PHP, so was not sure how to do it.

File: src/Instrumentation/CakePHP/composer.json [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  "description": "OpenTelemetry auto-instrumentation for CakePHP Framework.",
+  "keywords": ["opentelemetry", "otel", "open-telemetry", "tracing", "cakephp", "instrumentation"],
+  "type": "library",
+  "homepage": "https://opentelemetry.io/docs/php",
+  "readme": "./README.md",
+  "license": "Apache-2.0",
+  "minimum-stability": "dev",
+  "prefer-stable": true,
+  "require": {
+    "php": "^8.0",
```
Comment (brettmc) [inline src/Instrumentation/CakePHP/composer.json:12]: should also require cakephp?

File: src/Instrumentation/Slim/src/SlimInstrumentation.php
Comment (brettmc) [inline src/Instrumentation/Slim/src/SlimInstrumentation.php:1]: I think this has broken linting for Slim. In any case, it should not be part of this PR.

File: src/Instrumentation/CakePHP/composer.json (hunk 3 above)
Comment (AA2512) [reply src/Instrumentation/CakePHP/composer.json:12]: Yeah, will add that

File: src/Instrumentation/Slim/src/SlimInstrumentation.php
Comment (AA2512) [reply src/Instrumentation/Slim/src/SlimInstrumentation.php:1]: I just copied the code from main branch and updated it. But still not sure why is it showing up in the diff

Comment (AA2512) [reply src/Instrumentation/Slim/src/SlimInstrumentation.php:1]: Done

File: src/Instrumentation/CakePHP/src/CakePHPInstrumentation.php [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                    $span->setAttribute(TraceAttributes::NETWORK_PROTOCOL_VERSION, $response->getProtocolVersion());
+                    $span->setAttribute(TraceAttributes::HTTP_RESPONSE_BODY_SIZE, $response->getHeaderLine('Content-Length'));
+                }
+
+                $span->end();
+            }
+            
+        );
+       
+    }
```
Comment (ChrisLightfootWild) [inline src/Instrumentation/CakePHP/src/CakePHPInstrumentation.php:88]:
```suggestion
                }

                $span->end();
            },
        );
    }
```

**Review by ChrisLightfootWild — COMMENTED** [review]
Hey @AA2512, thanks for contributing :partying_face:

This PR was mentioned in the PHP-SIG today. To summarise:

 - an issue can be [created](https://github.com/open-telemetry/opentelemetry-php/issues/new?assignees=&labels=feature-request&projects=&template=feature_request.md) to mention the lack of tests, so someone else might help contribute those in future :crossed_fingers: 
 - a screenshot demonstrating this working would be beneficial, to prove that the hook is working as expected

I have left a couple of minor additional suggestions based on things I missed on my initial pass.

File: src/Instrumentation/CakePHP/src/CakePHPInstrumentation.php [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+                        ->setAttribute(TraceAttributes::URL_FULL, $request->getUri()->__toString())
+                        ->setAttribute(TraceAttributes::HTTP_REQUEST_METHOD, $request->getMethod())
+                        ->setAttribute(TraceAttributes::HTTP_REQUEST_BODY_SIZE, $request->getHeaderLine('Content-Length'))
+                        ->setAttribute(TraceAttributes::USER_AGENT_ORIGINAL, $request->getHeaderLine('User-Agent'))
+                        ->setAttribute(TraceAttributes::SERVER_ADDRESS, $request->getUri()->getHost())
+                        ->setAttribute(TraceAttributes::SERVER_PORT, $request->getUri()->getPort())
+                        ->setAttribute(TraceAttributes::URL_SCHEME, $request->getUri()->getScheme())
+                        ->setAttribute(TraceAttributes::URL_PATH, $request->getUri()->getPath())
+                        ->startSpan();
+                    $request = $request->withAttribute(SpanInterface::class, $span);
```
Comment (agoallikmaa) [inline src/Instrumentation/CakePHP/src/CakePHPInstrumentation.php:55]: This creates a new request object with that additional attribute. However, this new object is not passed on to anywhere, so it has no effect. This attribute is only needed if another hook uses it anyway, which does not seem to be the case here, so can just omit this line.

**Review by agoallikmaa — CHANGES_REQUESTED** [review]

File: src/Instrumentation/CakePHP/_register.php [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+use OpenTelemetry\Contrib\Instrumentation\CakePHP\CakePHPInstrumentation;
+use OpenTelemetry\SDK\Sdk;
+
+if (class_exists(Sdk::class) && Sdk::isInstrumentationDisabled(CakePHPInstrumentation::NAME) === true) {
+    return;
+}
+
+if (extension_loaded('opentelemetry') === false) {
+    trigger_error('The opentelemetry extension must be loaded in order to autoload the OpenTelemetry Cake PHP Framework auto-instrumentation', E_USER_WARNING);
```
[the remark this answers is not in the document]
Comment (AA2512) [reply src/Instrumentation/CakePHP/_register.php]: Done

File: src/Instrumentation/CakePHP/src/CakePHPInstrumentation.php (hunk 4 above)
Comment (AA2512) [reply src/Instrumentation/CakePHP/src/CakePHPInstrumentation.php]: Done

File: src/Instrumentation/CakePHP/src/CakePHPInstrumentation.php (hunk 5 above)
Comment (AA2512) [reply src/Instrumentation/CakePHP/src/CakePHPInstrumentation.php]: Done

File: src/Instrumentation/CakePHP/_register.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (AA2512) [reply src/Instrumentation/CakePHP/_register.php]: Screentshots: 
Screenshot 2024-04-17 at 10 23 33 AM
Screenshot 2024-04-17 at 10 24 04 AM

File: src/Instrumentation/CakePHP/composer.json [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    "ext-opentelemetry": "*",
+    "cakephp/cakephp": ">=4.0",
+    "ext-reflection": "*",
```
Comment (brettmc) [inline src/Instrumentation/CakePHP/composer.json:15]: I don't see this extension being used. If not required, it can be removed.

**Comment by brettmc** [comment]
Can you please add this to the top-level `.gitsplit.yaml` and also to `.github` like other packages do? Then linting, static analysis etc will run for this package (and we might see some new failures).

File: src/Instrumentation/CakePHP/_register.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (ChrisLightfootWild) [reply src/Instrumentation/CakePHP/_register.php]: Thank you! Based on the screenshot, it looks like the body size is not necessarily always present.

Might possibly be worth changing `$request->getHeaderLine('Content-Length')` to `$request->getHeaderLine('Content-Length') ?: null` when you are setting the attribute. Or omitting it completely if not set.

[the remark this answers is not in the document]
Comment (AA2512) [reply src/Instrumentation/CakePHP/_register.php]: Done

File: .github/workflows/php.yml [hunk 8]
```diff
@@ -90,6 +91,8 @@ jobs:
             php-version: 7.4
           - project: 'Instrumentation/OpenAIPHP'
             php-version: 8.0
+          - project: 'Instrumentation/CakePHP'
+            php-version: 8.0
```
Comment (brettmc) [inline .github/workflows/php.yml:95]: this excludes running tests against php 8.0, but composer.json says it supports 8.0. Did you mean to exclude 7.4?

Comment (AA2512) [reply .github/workflows/php.yml:95]: Yeah sorry. Should i exclude this entirely from here ? As we don't have any test cases setup for this currently ?

Comment (brettmc) [reply .github/workflows/php.yml:95]: No, it should still run even with zero tests (because there are a lot of other checks that also run that are important).

**Comment by brettmc** [comment]
@AA2512 can you take a look at the conflict and try to resolve it please?

File: .github/workflows/php.yml (hunk 8 above)
Comment (AA2512) [reply .github/workflows/php.yml]: Done

File: src/Instrumentation/CakePHP/src/CakePHPInstrumentation.php [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+                    ->setSpanKind(SpanKind::KIND_SERVER)
+                    ->setAttribute(TraceAttributes::CODE_FUNCTION, $function)
+                    ->setAttribute(TraceAttributes::CODE_NAMESPACE, $class)
+                    ->setAttribute(TraceAttributes::CODE_FILEPATH, $filename)
+                    ->setAttribute(TraceAttributes::CODE_LINENO, $lineno);
+                if ($request) {
+                    $parent = Globals::propagator()->extract($request->getHeaders());
+                    $span = $builder
+                        ->setParent($parent)
+                        ->setAttribute(TraceAttributes::URL_FULL, $request->getUri()->__toString())
```
Comment (ChrisLightfootWild) [inline src/Instrumentation/CakePHP/src/CakePHPInstrumentation.php:46]:
```suggestion
                        ->setAttribute(TraceAttributes::URL_FULL, (string) $request->getUri())
```

**Review by agoallikmaa — APPROVED** [review]

**Review by agoallikmaa — COMMENTED** [review]
Looks good, waiting for conflict resolution.

**Comment by AA2512** [comment]
Resolved the conflict.

**Comment by brettmc** [comment]
You need to remove 7.4 from the github config, and fix linting issues. If you follow our development recommendations, you can use make to run all of the checks that github actions will.

**Comment by AA2512** [comment]
How to run the php-cs-fixer as when i run the command 
PROJECT=CakePHP make all

I get the error:

docker-compose run --rm -w /usr/src/myapp/src/CakePHP php env XDEBUG_MODE=off vendor/bin/php-cs-fixer fix --config=.php-cs-fixer.php --using-cache=no -vvv
env: can't execute 'vendor/bin/php-cs-fixer': No such file or directory
make: *** [style] Error 127

**Comment by brettmc** [comment]
> How to run the php-cs-fixer as when i run the command

`PHP_VERSION=8.1 PROJECT=Instrumentation/CakePHP make all`

**Comment by brettmc** [comment]
@AA2512 it's looking good. To fix the failing tests, you can just add a `Unit` and `Integration` dir, with a `.gitkeep` file in it and commit that. It should be happy then.

**Comment by AA2512** [comment]
Hi, I'm not sure why it's failing for PHP 8.1 onwards. The error indicates some error intersection related error in psalm

**Comment by brettmc** [comment]
> Hi, I'm not sure why it's failing for PHP 8.1 onwards. The error indicates some error intersection related error in psalm

psalm looks like it is crashing on 8.0
does this not happen when you run psalm in our developer image? Or are you developing some other way?

**Comment by AA2512** [comment]
I didn't check after fixing the errors which it was throwing in the code. I just ran it, it throws the same error in the developer image too

**Comment by AA2512** [comment]
I think the issue was with the psalm version.
I have updated them. Let's see

File: src/Instrumentation/CakePHP/tests/Integration/CakeInstrumentationTest.php
Comment (brettmc) [reply src/Instrumentation/CakePHP/tests/Integration/CakeInstrumentationTest.php:1]: Agreed in SIG that we can accept without tests.

**Review by brettmc — APPROVED** [review]

**Comment by brettmc** [comment]
@AA2512 it's available to install via composer now: https://packagist.org/packages/open-telemetry/opentelemetry-auto-cakephp
Please test it out!

**Comment by AA2512** [comment]
Yep it's working fine.
